### ISSUE TITLE
[bug] unrelated_mode is not transitive

### ISSUE THREAD

**Issue by michaelmayer** [conan-io/conan#16366]
### Describe the bug

If I have a dependency to a package in "unrelated_mode" I would expect that all its transitive dependencies are also in "unrelated_mode".

However, they are still listed as requirement in my top level package. So, in order to "get rid" of those sub-dependencies, I have to manually put them into unrelated mode as well in the package_id method.

Don't know if that's really a bug or if this behaviour is somehow expected or necessary. It surely surprised me, though (because those sub-dependencies are OS dependent, in my case).

### How to reproduce it

Just set up a simple example project in three parts (top, middle, last):
```sh
$ mkdir last && cd last
$ conan new cmake_lib -d name=last -d version=1.0
$ conan create -s build_type=Release .
$ cd ..
$ mkdir middle && cd middle
$ conan new cmake_lib -d name=middle -d requires=last/1.0 -d version=1.0
$ conan create -s build_type=Release .
$ cd ..
$ mkdir top && cd top
$ conan new cmake_exe -d name=top -d requires=middle/1.0 -d version=1.0
$ <edit> conanfile.py adding 'package_id_mode="unrelated_mode"' to the self.requires() command
$ conan create -s build_type=Release .
$ conan list top/1.0:*
Local Cache
top
      [...]
              requires
                last/1.0#13bc058f4572b233101570340f9e8438:4d31800c65d4e13abdbe8ca04d2e1d94503a9c17
```
As expected, the `middle` package is gone from our requirements. However, the `last` is still here, although I put my _direct_ requirement to `middle` already into "unrelated_mode".

**Comment by memsharded** [comment]
Hi @michaelmayer 

I'd say this is expected. The ``self.requires("somedep", package_id_mode="unrelated")`` model is not expected to be transitive, because in general it cannot be deduced this transitivity is correct. There are tons of cases where this is not the case, for example, when you depend on some "aggregator" package or header-only-like package that doesn't affect directly to the ``package_id`` of the consumer application, but such "aggregator" does in turn depends on several regular static-libraries.

As you guessed, one of the mechanisms to force the explicit erasure of the influence of ``last`` is to explicitly change it in the ``package_id()`` method, not via the ``self.requires``, because the requires only affect at the directly required one.

However, I'd probably recommend having a look to other mechanisms. If the ``last`` package is actually special, and you know how it works for its consumers already, why not defining its default mode in the ``last`` recipe ``conanfile.py``? Have you had a look to:
https://docs.conan.io/2/reference/binary_model/custom_compatibility.html#custom-package-id-modes-for-recipe-consumers ?

**Comment by michaelmayer** [comment]
Hi @memsharded 

Thanks for your quick response! However, I'm not convinced, yet 😄 

> If the last package is actually special, 

Well, the `last` package is not special at all. My usage of the `middle` package is. The `top` package doesn't even know the the `last` package is used, it doesn't want to know.

In my real use-case, the reason for the `top` (as consumer) to mark `middle` as "unrelated" is, that it's actually not linked into the (shared) library, but only used by some example code which is packaged with it.  
Furthermore, this requirement can be switched off by some package options, hence the need to get rid of it, but only for the package_id. The requirement still works as expected when the package is consumed with the "examples" option switched on.

> There are tons of cases where this is not the case, for example, when you depend on some "aggregator" package or header-only-like package that doesn't affect directly to the package_id of the consumer application, but such "aggregator" does in turn depends on several regular static-libraries.

Hmm, if I do rely on an aggregator package which depends on other libs, and those libs actually contribute to my package, then I, as a consumer, wouldn't (or shouldn't) mark the aggregator package as "unrelated".  
When I, as a consumer, on the other hand, mark a package as unrelated, of course I don't want any of its sub-dependencies to "pollute" my package_id. But of course that's only my POV.

In our case we have an internal package, used only for said example code, which changed, and now consumes the [util-linux-libuuid](https://github.com/conan-io/conan-center-index/tree/master/recipes/util-linux-libuuid). After fixing the recipe to also ignore the libuuid, it seemed to work again ... until we tried to build for macos. Well, because of:
```
        if self.settings.os == "Macos":
            self.requires("libgettext/0.22")
```

But, I'm pretty sure you have good reasons (still) not to implement the "unrelated_mode" transitively. Just wanted to contribute my thoughts and experience. We can perfectly live with it, as is. Feel free to close this issue.

Thanks for your hard work!

**Comment by memsharded** [comment]
> Thanks for your quick response! However, I'm not convinced, yet 😄

Fair, let me try again 😄 

The problem is less evident, and I agree with some of your observations, with the ``unrelated_mode``, which is a bit kind of special.

The issue that we had is more subtle changes, for example, a user has ``libc -> libb -> liba``
And then because of some details about how ``libb`` developers use versions, the ``libc`` developer decides to do:

```python
name = "libc"

def requirements(self):
    self.requires("libb/0.1", package_id_mode="patch_mode")
```
instead of using the Conan defined defaults or Conan configured (via ``conf`` for example).

Now, if we propagate upstream this ``patch_mode`` definition:
- It will also make ``libc`` depend on ``liba`` in ``patch_mode``, but the versioning of ``liba`` was correct in the first place, and now it will require to do way more rebuilds from source of ``libc`` that are not necessary.
- The propagation will also override the "embed" and "non_embed" modes. Conan can manage correctly when the libraries are static or shared or a combination of both, and produce correct binaries with the minimal necessary rebuild, but with the override of ``patch_mode`` upstream now we have:
  - For all static libraries and all shared libraries, the number of unnecessary rebuilds will increase
  - For a mixture of shared libraries and static libraries, the builds will not be enough, for example if ``libb`` is a shared library and ``liba`` is a static library, and ``libb`` is only rebuilt when patches change, but not when there are new revisions of ``liba``, then ``libb`` will be broken for every new revision of ``liba``.

Also note that the propagation of this upstream will imply that when you are building ``libb`` in isolation, that is, not as a dependency of ``libc``, but by its own, its ``package_id_mode`` will not be overriden (there is no one downstream to override), so it will use a different ``package_id_mode`` to build its binary. Then, when ``libc`` is going to be built against this pre-compiled ``libb``, then it will fail saying that "Missing binary for libb", because now it wants a different binary of ``libb`` than the one that was built in isolation.

So, making the ``self.requires(..., package_id_mode=xxx)`` non propagating upstream is in the majority of cases more expected and produces less bad surprises in the behavior.

Possibly the issue is that the ``self.requires(..., package_id_mode=xxx)`` does not explain well enough its use case. In general, it should be avoided, and possibly should be used only as a last resort and for very exceptional scenarios. In most cases:

- The Conan defaults are well thought, try to use them as much as possible.
- When some package is special, it can define its modes for consumers in its recipe. This is self-contained (does not suffer the problem above), and avoids the consumers to have to worry about these details.
- When necessary to do changes in the consumer side, the ``package_id()`` method will be more recommended, because there you can have full control over the dependencies, if they are shared or static, etc, and act accordingly

**Comment by michaelmayer** [comment]
Ah yes, my train of thoughts is only valid for the "unrelated_mode", true. Thanks for the detailed explanation!

Would it maybe help if I don't mark the requirement as unrelated, but instead remove it completely in the package_id method? Are their sub-dependencies then removed as well?

**Comment by memsharded** [comment]
> Ah yes, my train of thoughts is only valid for the "unrelated_mode", true. Thanks for the detailed explanation!

Yes, it makes more sense for the ``unrelated_mode`` than other modes, but it will be super challenging to have Conan propagate this based on the specific value, if it doesn't make the most sense for the general case of all modes.

> Would it maybe help if I don't mark the requirement as unrelated, but instead remove it completely in the package_id method? Are their sub-dependencies then removed as well?

No, unfortunately the sub-dependencies will not be automatically removed, you need to remove them yourself. Does the package have many other dependencies?


Maybe another strategy would be to put the dependency as ``test_requires``?. Note that "examples" in your case is really close to "tests" in that regard. So if the examples are packaged, but the transitive dependencies are not needed at all for those examples, then putting them as ``test_requires`` will be even better, because they also won't cause conflicts downstream, it is not only about the ``package_id`` effect, but also about propagation downstream of those dependencies. By default ``test_requires`` do not affect the ``package_id``, neither the direct or transitive dependencies of ``test_requires``.

**Comment by michaelmayer** [comment]
> Does the package have many other dependencies?

It's not so much about the number of dependencies. It's more about stability. As soon as the dependency changes its sub-dependencies (sometimes without our knowledge), we have to update our top recipe.

Do I understand this correctly, that in the `package_id` method we don't have access to the requirements graph (or tree) anymore, but we just have a flat list of them?

> Maybe another strategy would be to put the dependency as test_requires?

Under "normal" circumstances, yes, I agree totally. Unfortunately, consumers of the package need to be able to build the examples, and therefor also need those requirements in that case. Specifically, on the CI, we retrieve the package with `conan install --requires=xxx ...` and then build the examples. AFAIK, test- & tool- requirements are not accessible for the consumer in any way, are they?  
Note, that in a perfect world those "examples" would of course reside in a different conan package and just rely on the shared library package. But it's just super convenient for us devs to have them in the same repo (yes, I know about the "editable" command 😄).

> it is not only about the package_id effect, but also about propagation downstream of those dependencies.

Yes, that's exactly why we have those requirements only optionally.

**Comment by memsharded** [comment]
> Do I understand this correctly, that in the package_id method we don't have access to the requirements graph (or tree) anymore, but we just have a flat list of them?

Only the ``self.info.requires`` is flat, but via the ``self.dependencies`` interface you have access to the graph:
- You can filter in ``self.dependencies`` the direct dependencies.
- The ``dep.dependencies`` attribute for the dependencies is also public and available

So ``self.dependencies["mydep"].dependencies`` (it is a dict, you can get its ``.values()``, then the ``ref`` of each one) will give you the packages that you want, so I think that putting all transitive dependencies of ``mydep`` to ``unrelated_mode`` should be just a couple of lines. Please let me know if you need help with this.

**Comment by michaelmayer** [comment]
Ah, very cool. That actually works for us.

Thank you very much!

**Comment by memsharded** [comment]
> Ah, very cool. That actually works for us.

Excellent, thanks for the feedback.

> Thank you very much!

Very happy to help! 🙂
